File: projects/rpn2tex/migrations/go-module-by-module-1/token.go
package main

import "fmt"

// TokenType represents the type of a lexical token in the RPN expression.
type TokenType int

// Token types supported by the rpn2tex lexer.
const (
	NUMBER TokenType = iota
	PLUS
	MINUS
	MULT
	DIV
	EOF
)

// String returns the string representation of a TokenType.
func (tt TokenType) String() string {
	switch tt {
	case NUMBER:
		return "NUMBER"
	case PLUS:
		return "PLUS"
	case MINUS:
		return "MINUS"
	case MULT:
		return "MULT"
	case DIV:
		return "DIV"
	case EOF:
		return "EOF"
	default:
		return fmt.Sprintf("TokenType(%d)", tt)
	}
}

// Token represents a lexical token with position information.
// It is immutable by design - all fields are read-only after creation.
type Token struct {
	Type   TokenType
	Value  string
	Line   int
	Column int
}

// String returns the string representation of a Token for debugging.
// Format: Token(TYPE, "value", line:column)
func (t Token) String() string {
	return fmt.Sprintf("Token(%s, %q, %d:%d)", t.Type, t.Value, t.Line, t.Column)
}

File: projects/rpn2tex/migrations/java-module-by-module-3/src/main/java/com/rpn2tex/Token.java
package com.rpn2tex;

import java.util.Objects;

/**
 * Immutable lexical token representing a single element in an RPN expression.
 *
 * <p>A token consists of:
 * <ul>
 *   <li>A {@link TokenType} indicating the category of the token</li>
 *   <li>A string value containing the exact lexeme from the source</li>
 *   <li>A 1-based line number for error reporting</li>
 *   <li>A 1-based column number for error reporting</li>
 * </ul>
 *
 * <p>Position tracking enables precise error messages that point to the exact
 * location of issues in the source text. All position values use 1-based indexing
 * to match common editor conventions.
 *
 * <p><b>Examples:</b>
 * <pre>{@code
 * // Number token at position (1, 1) with value "5"
 * Token number = new Token(TokenType.NUMBER, "5", 1, 1);
 *
 * // Addition operator at position (1, 3)
 * Token plus = new Token(TokenType.PLUS, "+", 1, 3);
 *
 * // EOF token at end of input
 * Token eof = new Token(TokenType.EOF, "", 1, 10);
 * }</pre>
 *
 * @param type   the token type (must not be null)
 * @param value  the string value from the source (must not be null)
 * @param line   the 1-based line number where the token appears
 * @param column the 1-based column number where the token starts
 * @since 1.0.0
 */
public record Token(
    TokenType type,
    String value,
    int line,
    int column
) {
    /**
     * Compact constructor with validation.
     *
     * @throws NullPointerException if type or value is null
     */
    public Token {
        Objects.requireNonNull(type, "type must not be null");
        Objects.requireNonNull(value, "value must not be null");
    }

    /**
     * Returns a string representation of this token.
     *
     * <p>The format is: {@code Token(TYPE, "value", line:column)}
     *
     * <p><b>Example:</b> {@code Token(NUMBER, "5", 1:1)}
     *
     * @return a string representation of this token
     */
    @Override
    public String toString() {
        return String.format("Token(%s, \"%s\", %d:%d)", type, value, line, column);
    }
}
